lucet-runtime/lucet-runtime-internals/src/instance/execution.rs

```diff
@@ -125,6 +127,41 @@ pub struct KillState {
     tid_change_notifier: Condvar,
 }
 
+/// This function is called by `lucet_context_activate` before entering a region of guest code.
+///
+/// This function is responsible for setting the terminable flag on the instance's
+/// [`KillState`](struct.KillState.html), so that we can appropriately signal the instance to
+/// terminate. If a pending instance was signalled before it started executing guest code, we will
+/// terminate the instance and swap back to its host context.
+///
+///
+/// entering a guest region, so that we can use signals to terminate the instance.  check that the
+/// instance has not already been remotely terminated by a killswitch.  set the `terminable` flag
+/// on our instance's [`KillState`](struct.KillState.html).
+///
+/// # Safety
+///
+/// This function will call [Instance::terminate](struct.Instance.html#method.terminate) if the
+/// execution domain is [`Domain::Cancelled`](enum.Domain.html). In that case, this function will
+/// not return, and we will swap back to the host context without unwinding.
+pub unsafe extern "C" fn enter_guest_region(instance: *mut Instance) {
+    let mut current_domain = (*instance).kill_state.execution_domain.lock().unwrap();
+    match *current_domain {
+        // We are about to start executing a guest. Set the execution domain accordingly.
+        Domain::Pending => *current_domain = Domain::Guest,
+        // We are about to enter guest code for an instance that has already been terminated.
+        // Instead, we should forgo entering the guest region entirely, terminate the instance, and
+        // swap back to the host without any unwinding.
+        Domain::Cancelled => (*instance).terminate(TerminationDetails::Remote),
+        // If we are in some other domain, do nothing. These variants are handled elsewhere.
```

Comment:
I think it would in fact be an error to reach this point in any other domain:
* `Guest` would be a no-no because that suggests we might send a `SIGALRM` before we've indicated it safe to do so by `enable_termination`
* `Terminated` would be a no-no because this would be a zombie ("this has been stopped, but we're just starting again?")
* `Hostcall` would suggest that a host has called into guest code, which isn't bad on its own, but that should _not_ involve going through `enter_guest_region` a second time

so I think you're perfectly justified in making this arm a `panic`, since reaching this would be some form of implementation error.

Reply:
We talked about this out-of-band, but for posterity: Yes! This totally makes sense. All of these cases would mean that something has gone terribly wrong. I added panic statements here.